File: frontend/src/api/onboarding.ts
import { apiGet, apiPut, apiPost } from '@/api/client'
import type { WizardRoute } from '@/api/wizard'

export type { WizardRoute }

export interface OnboardingRouteStatus {
  route: WizardRoute | null
  demo_available: boolean
  demo_school_name: string
  has_demo_semester: boolean
  has_formal_semester: boolean
  can_reselect: boolean
  resume_step: number
  resume_semester_id: number | null
}

export interface OnboardingAction {
  stage: string
  label: string
  href: string
  blocking_reason?: string
}

export interface P0Stage {
  key: string
  label: string
  complete: boolean
  status: 'complete' | 'blocked' | 'pending'
  blocking_reason: string
  next_action: OnboardingAction | null
  details: Record<string, unknown>
}

export interface OnboardingStatus {
  first_success: boolean
  wizard_completed: boolean
  current_semester: {
    id: number
    label: string
    is_demo: boolean
  } | null
  stages: P0Stage[]
  p0_todos: P0Stage[]
  next_action: OnboardingAction | null
}

export const getOnboardingStatus = () => apiGet<OnboardingStatus>('/onboarding/status')
export const getOnboardingRoute = () => apiGet<OnboardingRouteStatus>('/onboarding/route')
export const chooseOnboardingRoute = (route: WizardRoute) =>
  apiPut<OnboardingRouteStatus>('/onboarding/route', { route })
// POST keeps the entry point usable for clients that submit a first-use form.
export const chooseOnboardingRoutePost = (route: WizardRoute) =>
  apiPost<OnboardingRouteStatus>('/onboarding/route', { route })

File: backend/app/schemas/onboarding.py
"""首次成功引导与 P0 待办读模型。"""

from typing import Literal

from pydantic import BaseModel, Field

from app.schemas.semester import SemesterListItem

StageStatus = Literal["complete", "blocked", "pending"]


class OnboardingAction(BaseModel):
    stage: str
    label: str
    href: str
    blocking_reason: str = ""


class P0Stage(BaseModel):
    key: str
    label: str
    complete: bool
    status: StageStatus
    blocking_reason: str = ""
    next_action: OnboardingAction | None = None
    details: dict[str, object] = Field(default_factory=dict)


class OnboardingStatusOut(BaseModel):
    """由当前业务数据实时推导的首次成功状态。"""

    first_success: bool
    wizard_completed: bool
    current_semester: SemesterListItem | None = None
    stages: list[P0Stage] = Field(default_factory=list)
    p0_todos: list[P0Stage] = Field(default_factory=list)
    next_action: OnboardingAction | None = None
